### PR TITLE
imp: check if there are unreviewed paths by METADATA files

### PR DESCRIPTION
**Pull request by Pawka**
Call custom API endpoint to validate if there are unreviewed paths by METADATA files on the repository. Ask user if he/she wants to continue if such paths exist.

### CODE REVIEW

**Comment by Pawka** [comment]
Example:

Screenshot from 2019-11-27 14-38-14

This flow fails on the first unreviewed path. We do not return all unreviewed paths to reduce traffic and keep performance high.

Also maybe it is worth to roll out this via opt-in flag?

File: src/workflow/ArcanistLandWorkflow.php [hunk 1]
```diff
@@ -942,6 +942,82 @@ private function findRevision() {
         pht('Accepted'));
     }
 
+    // Check if all paths were reviewed by reviewers listed on METADATA files.
```
Comment (artms) [inline src/workflow/ArcanistLandWorkflow.php]: Add // UBER CODE?

File: src/workflow/ArcanistLandWorkflow.php [hunk 2]
```diff
@@ -942,6 +942,82 @@ private function findRevision() {
...
+        'reviewers defined on METADATA files. This validation will be '.
+        'performed during `git push` operation.'
+      );
+      id(new PhutilConsoleBlock())
+        ->addParagraph(tsprintf(
+          '<bg:yellow>** %s **</bg>', 'WARNING'))
+        ->addParagraph(tsprintf('%B', $warning))
+        ->draw();
+    }
+
```
Comment (artms) [inline src/workflow/ArcanistLandWorkflow.php]: What will happen if repository doesn't have METADATA configured? Maybe we should check tag and skip warning if repository is not using METADATA?

File: src/workflow/ArcanistLandWorkflow.php [hunk 3]
```diff
@@ -942,6 +942,82 @@ private function findRevision() {
...
+      if (array_key_exists('pass', $uber_metadata_unreviewed_paths)
+        && false === $uber_metadata_unreviewed_paths['pass']
+      ) {
+
+        id(new PhutilConsoleWarning('WARNING', pht(
+          'Revision contains paths which were not reviewed by METADATA '.
+          'reviewers. Likely land operation will be blocked.'
+        )))->draw();
+
+        foreach ($uber_metadata_unreviewed_paths['groups'] as $group) {
```
Comment (artms) [inline src/workflow/ArcanistLandWorkflow.php]: What about `users`?

File: src/workflow/ArcanistLandWorkflow.php [hunk 4]
```diff
@@ -942,6 +942,82 @@ private function findRevision() {
...
+              }, $group['reviewers'][0]['groups']);
+            }
+            if (!empty($group['reviewers'][0]['users'])) {
+              $users = array_map(function($v) {
+                return '@'.$v;
+              }, $group['reviewers'][0]['users']);
+              $reviewers = array_merge($reviewers, $users);
+            }
+          }
+          $suggestion = pht('No suggested reviewers found');
```
Comment (artms) [inline src/workflow/ArcanistLandWorkflow.php]: 'No suggested reviewers were found'

File: src/workflow/ArcanistLandWorkflow.php [hunk 5]
```diff
@@ -942,6 +942,82 @@ private function findRevision() {
...
 
+    // Check if all paths were reviewed by reviewers listed on METADATA files.
+    // If this check throws an exception - silently pass.
+    try {
+      $uber_metadata_unreviewed_paths = $this->getConduit()->callMethodSynchronous(
+        'uber_metadata.unreviewed_paths',
+        array(
+          'revisionid' => $rev_id,
+        ));
+      if (array_key_exists('pass', $uber_metadata_unreviewed_paths)
```
Comment (artms) [inline src/workflow/ArcanistLandWorkflow.php]: At 966 line you use `isset($group['reviewers'][0]` to check key presence, looks like a bit inconsistent

File: src/workflow/ArcanistLandWorkflow.php [hunk 6]
```diff
@@ -942,6 +942,82 @@ private function findRevision() {
...
+          'Revision contains paths which were not reviewed by METADATA '.
+          'reviewers. Likely land operation will be blocked.'
+        )))->draw();
+
+        foreach ($uber_metadata_unreviewed_paths['groups'] as $group) {
+
+          $block = new PhutilConsoleBlock();
+          $reviewers = [];
+          if (isset($group['reviewers'][0])) {
+            if (!empty($group['reviewers'][0]['groups'])) {
```
Comment (artms) [inline src/workflow/ArcanistLandWorkflow.php]: Could potentially be replaced with single `array_map(..., idx($group['reviewers'][0], "groups", array()))?`

File: src/workflow/ArcanistLandWorkflow.php (hunk 2 above)
Comment (Pawka) [reply src/workflow/ArcanistLandWorkflow.php]: This check is performed on API side and `"pass":true` is returned all the time.

File: src/workflow/ArcanistLandWorkflow.php (hunk 5 above)
Comment (Pawka) [reply src/workflow/ArcanistLandWorkflow.php]: `isset` and `array_key_exists` behavior is different. `isset()` does not return TRUE for array keys that correspond to a NULL value, while `array_key_exists()` does. I wanted to be very precise of "pass" check while the "reviewers" array is just a helper to print information to the user.

File: src/workflow/ArcanistLandWorkflow.php (hunk 3 above)
Comment (Pawka) [reply src/workflow/ArcanistLandWorkflow.php]: This is "reviewers group" array which contains paths and reviewers. Meanwhile reviewers contains ldap groups and users. I've added JSON example to the comment to make this flow more visible.

File: src/workflow/ArcanistLandWorkflow.php (hunk 6 above)
Comment (Pawka) [reply src/workflow/ArcanistLandWorkflow.php]: Potentially yes but that doesn't give any additional value.

**Pull request merged**